File: leetcode/946.验证栈序列.py
#
# @lc app=leetcode.cn id=946 lang=python3
#
# [946] 验证栈序列
#
from typing import List


# @lc code=start
class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        stack = []
        cache_set = set()
        i = 0
        n = len(popped)
        for num in pushed:
            stack.append(num)
            cache_set.add(num)
            while i < n and popped[i] in cache_set:
                if popped[i] != stack[-1]:
                    return False
                stack.pop()
                i += 1
        return True


# @lc code=end
print(Solution().validateStackSequences(
    pushed=[1, 2, 3, 4, 5], popped=[4, 5, 3, 2, 1]
))
print(Solution().validateStackSequences(
    pushed=[1, 2, 3, 4, 5], popped=[4, 3, 5, 1, 2]
))

File: leetcode/946.验证栈序列.cpp
#include <iostream>
#include <vector>
#include <set>
using namespace std;
/*
 * @lc app=leetcode.cn id=946 lang=cpp
 *
 * [946] 验证栈序列
 */

// @lc code=start
class Solution
{
public:
    bool validateStackSequences(vector<int> &pushed, vector<int> &popped)
    {
        vector<int> stack;
        set<int> cache_set;
        int i = 0;
        int n = popped.size();
        for (auto &&num : pushed)
        {
            stack.push_back(num);
            cache_set.emplace(num);
            while (i < n && cache_set.count(popped[i]))
            {
                if (popped[i] != stack.back())
                    return false;
                stack.pop_back();
                i++;
            }
        }
        return true;
    }
};
// @lc code=end
